models/utils.py

```diff
@@ -0,0 +1,140 @@
+        i,
+        commission_rate,
+        self_stake_base,
+        self_stake_max,
+        engagement_base,
+        engagement_max,
+        delegation,
+    ) for i in range(n)]
+    validators.reverse()
+
+    if shuffle_engagements:
+        engagements = [validator['engagement'] for validator in validators]
+        random.shuffle(engagements)
+        for i, validator in enumerate(validators):
+            validator['engagement'] = engagements[i]
+
+    return validators
+
+
+def create_delegator(delegation, validator_index=None):
+    return {
+        'delegation': delegation,
+        'validator': validator_index,
+    }
+
+
+def create_delegators(n_delegators, validators, delegation_stake_multiple):
+    sum_validators_self_stake = sum(
+        [validator['self_stake'] for validator in validators])
+    delegation = delegation_stake_multiple * \
+        sum_validators_self_stake / n_delegators
+    return [create_delegator(delegation) for _ in range(n_delegators)]
+
+
+def calculate_voting_weight(validator, params):
+    consensus_algorithm = params['consensus_algorithm']
+
+    if consensus_algorithm == 'PoE':
+        return poe.calculate_voting_weight(validator, params)
+    if consensus_algorithm == 'PoS':
+        return pos.calculate_voting_weight(validator, params)
+
+    raise ValueError('Consensus algorithm not known')
+
+
+def calculate_total_voting_weight(validators, params):
+    return sum([calculate_voting_weight(validator, params) for validator in validators])
+
+
+def calculate_reward(validator, total_voting_weight, params):
+    consensus_algorithm = params['consensus_algorithm']
+
+    if consensus_algorithm == 'PoE':
+        return poe.calculate_reward(validator, total_voting_weight, params)
+    if consensus_algorithm == 'PoS':
+        return pos.calculate_reward(validator, total_voting_weight, params)
+
+    raise ValueError('Consensus algorithm not known')
+
+
+def calculate_yield(validator, total_voting_weight, params):
+    consensus_algorithm = params['consensus_algorithm']
+
+    if consensus_algorithm == 'PoE':
+        return poe.calculate_yield(validator, total_voting_weight, params)
+    if consensus_algorithm == 'PoS':
+        return pos.calculate_yield(validator, total_voting_weight, params)
+
+    raise ValueError('Consensus algorithm not known')
+
+
+def hypothesize_yield(delegator, validator_index, validators, params):
+    current_validator_index = delegator['validator']
+    delegation = delegator['delegation']
+
+    if current_validator_index == validator_index:
+        total_voting_weight = calculate_total_voting_weight(
+            validators, params)
+        return calculate_yield(validators[current_validator_index], total_voting_weight, params)
+
+    if current_validator_index != None:
+        validators[current_validator_index]['delegation'] -= delegation
+
+    next_validator = validators[validator_index]
+    next_validator['delegation'] += delegation
```

Comment:
And here.

Doesn't this make the next calculations run on the data incorrect?
Maybe copy `validators` before doing the manipulation?

Reply:
`validators` is a `copy.deepcopy` passed into this function. Maybe it makes more sense to move it inside.

Reply:
Yeah, best to keep the safety assumptions inside the code rather than assume them from outside (except in languages where the compiler can enforce such stuff)